src/plugins/WebOfTrust/util/SynchronousDelayedBackgroundJob.java

```diff
@@ -0,0 +1,231 @@
+        // deadlock.
+        if (runningJobThread == Thread.currentThread()) {
+            return;
+        }
+        // Wait for the relevant execution to complete.
+        long waitForRunCount;
+        if (runningJobThread == null) {
+            // Wait for next run to finish
+            waitForRunCount = runCount + 1;
+        } else {
+            // Wait for next run after current run to finish
+            waitForRunCount = runCount + 2;
+        }
+        while (runCount < waitForRunCount) {
+            try {
+                wait();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                return;
+            }
+        }
+    }
+
+    /**
+     * Implementation of {@link #triggerExecution(long)} that does everything expected from that
+     * method, except it does not check for termination (the caller is assumed to have done this)
+     * and does not wait for the execution to complete.
+     * @see #triggerExecution(long)
+     * Caller must ensure synchronization on {@code this}.
+     */
+    private void triggerExecutionAsynchronously(long delay) {
+        if (delay < 0) {
+            delay = 0;
+        }
+        long newDeadline = System.currentTimeMillis() + delay;
+        if (newDeadline < nextRunDeadline) {
+            nextRunDeadline = newDeadline;
+            if (!nextRunScheduled) {
+                // Create a new waiting job thread using the given delay.
+                nextRunScheduled = true;
+                startNewJobThread();
+            } else {
+                // Notify waiting job thread that it should start earlier.
+                notifyAll();
+            }
+        }
+    }
+
+    /**
+     * Starts a new job daemon thread that:
+     * 1. Waits for the deadline to expire
+     * 2. Waits until the previous job is done, if any
+     * 3. Resets {@link #nextRunDeadline} / {@link #nextRunScheduled}
+     * 4. Runs the background job
+     * 5. Increments {@link #runCount}
+     * 6. Notifies all threads waiting upon its completion
+     * Caller must ensure synchronization on {@code this}.
+     */
+    private void startNewJobThread() {
+        Thread jobThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                synchronized(SynchronousDelayedBackgroundJob.this) {
+                    long delay;
+                    // 1. Wait for the deadline to expire
+                    while ((delay = nextRunDeadline - System.currentTimeMillis()) > 0) {
+                        try {
+                            SynchronousDelayedBackgroundJob.this.wait(delay);
+                        } catch (InterruptedException e) {
+                            // Ignore. We can't really return here, since there is at
+                            // least one thread waiting for this thread to complete
+                            // (possibly including side effects caused by its run).
```

Comment:
Well I can only speak for WOT: I don't *need* a triggerExecution() to guarantee execution to happen if a shutdown comes in between. If the work was that important, it wouldn't delay it by using a background job.
So I would be OK with relaxing the contract to cause scheduled executions to be ignored if a shutdown happens before.
However, as I won't use the class outside of unit tests, this is not a blocker I think. You might only want to do it if it simplifies the planned shutdown issue fixes in the previous comment block.